src/core/src/service_interfaces/StatusHandler.py

```diff
@@ -812,8 +832,164 @@ def __set_errors_json(self, error_count_by_operation, errors_by_operation):
+
+        if len(min_assessment_patches_to_retain) > 0:
+            max_package_list_capacity = max_package_list_capacity - self.__calc_package_payload_size_on_disk(min_assessment_patches_to_retain)
+
+        # Apply high priority (Failed, Installed) and low priority (Pending, Excluded, Not_Selected) installation logic, and keep min 5 assessment packages
+        if low_pri_index is not None:
+            installation_high_pri = installation_packages[:low_pri_index]
+            installation_low_pri = installation_packages[low_pri_index:]
+
+        packages_retained_in_install_high_pri, packages_removed_from_install_high_pri, remaining_capacity_for_truncation = self.__apply_truncation(installation_high_pri, max_package_list_capacity)
+        packages_retained_in_assessment, packages_removed_from_assessment, remaining_capacity_for_truncation = self.__apply_truncation(remaining_assessment_patches, remaining_capacity_for_truncation)
+        packages_retained_in_install_low_pri, packages_removed_from_install_low_pri, remaining_capacity_for_truncation = self.__apply_truncation(installation_low_pri, remaining_capacity_for_truncation)
+        self.composite_logger.log_debug("After truncating packages, the remaining max capacity for truncating packages is {0}".format(remaining_capacity_for_truncation))
+
+        truncated_installation_list = packages_retained_in_install_high_pri + packages_retained_in_install_low_pri
+        packages_removed_from_installation = packages_removed_from_install_high_pri + packages_removed_from_install_low_pri
+        truncated_assessment_list = min_assessment_patches_to_retain + packages_retained_in_assessment
+
+        return truncated_assessment_list, packages_removed_from_assessment, truncated_installation_list, packages_removed_from_installation
+
+    def __get_installation_low_pri_index(self, priority_sorted_installation_packages):
+        """" Get the first index of Pending, Excluded, or Not_Selected from installation packages """
+        for index, package in enumerate(priority_sorted_installation_packages):
+            package_state = package['patchInstallationState']
+            if Constants.PENDING in package_state or Constants.EXCLUDED in package_state or Constants.NOT_SELECTED in package_state:
+                return index
+
+        return None
+
+    def __apply_truncation(self, package_list, capacity):
+        """ Binary search
+        Instead of checking list[middel_index] >= target, check byte_size(list[:middle_index]),
+        as byte_size[list[:i]] is monotonically increasing, i.e.
+        byte_size[list[:1]] < byte_size[list[:2]] < byte_size[list[:3]] ...
+        return truncated_list, packages_removed_from_list, and remaining max_package_list_capacity
+        """
+        left_index = 0
+        right_index = len(package_list) - 1
+
+        # Empty list after 2xjson.dumps have 4-5 bytes, no truncation, keep list capacity as it is
```

Comment:
Not very clear what this means. Rephrase

Reply:
if there's empty list, we won't perform truncation on it. comments rephrase, done